### PR TITLE
Enemy Ball Placement Play

### LINKED ISSUE #1999: Add Enemy Ball Placement Play
### Description of the task
According to the [rules](https://robocup-ssl.github.io/ssl-rules/sslrules.html#_ball_placement_interference), we need to stay away from the line from the ball to the ball placement point. We need a play to do that. It should be applicable when the gamestate is theirballplacement and we should also update stp_refbox_game_state_play_selection_test with ball_placement_them when this play is added.

### Acceptance criteria

- [ ] add enemy ball placement play
- [ ] simulated play test

### PR DESCRIPTION
**Pull request by PhilD71**
### Description
A play to be run during Enemy Ball Placement. The idea is to use the time during ball placement to transition our robots into a neutral state. This allows us to be ready for whatever play comes next.

### Testing Done
Simulated Testing [Not complete]
- need to add more tests

Unit Testing [Not started]
- need to add unit tests

### Resolved Issues
resolves #1999 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/geom/polygon.cpp [hunk 1]
```diff
@@ -50,6 +50,12 @@ Point Polygon::centroid() const
 }
 
 Polygon Polygon::expand(const Vector& expansion_vector) const
+{
+    Point centroid_point = centroid();
+    return expand(expansion_vector, centroid_point);
+}
+
+Polygon Polygon::expand(const Vector& expansion_vector, Point centroid_point) const
```
Comment (jonathanlew) [inline src/software/geom/polygon.cpp]: we can get rid of this old expand function entirely

File: src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp [hunk 2]
```diff
@@ -199,3 +200,46 @@ ObstaclePtr RobotNavigationObstacleFactory::createFromFieldRectangle(
...
+     *           +-------+-------+
+     *           |               |
+     *           |               |
+     *           +-------+-------+
+     *        place_l          place_r
+     */
+    const double RADIUS = 0.5;
+
+    Vector place_to_ball = placement_point.toVector() - ball_point.toVector();
+    Vector ball_to_place = ball_point.toVector() - placement_point.toVector();
```
Comment (jonathanlew) [inline src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp]: these are opposite vectors, so we can simplify the code by just calculating one of them and flipping +/-

File: src/software/ai/motion_constraint/motion_constraint.h [hunk 3]
```diff
@@ -24,4 +24,6 @@ MAKE_ENUM(MotionConstraint,
           // The enemy half of the field
           ENEMY_HALF,
           // The friendly half of the field
-          FRIENDLY_HALF);
+          FRIENDLY_HALF,
+          // Path between ball and placement point in ball placement
+          BALL_PLACEMENT_ZONE);
```
Comment (jonathanlew) [inline src/software/ai/motion_constraint/motion_constraint.h]: nit: I think something like AVOID_BALL_PLACEMENT_INTERFERENCE is more precise

File: src/software/geom/polygon.h [hunk 4]
```diff
@@ -40,8 +40,19 @@ class Polygon : public virtual Shape
...
 
+    /**
+     * Returns the Polygon expanded in all directions by the expansion_amount
+     *
+     * @param expansion_amount a non-negative expansion amount
+     * @throw std::invalid_argument if expansion_amount is negative
+     *
+     * @return a Polygon expanded in the direction of expansion_vector
+     */
+    Polygon expand(double expansion_amount) const;
```
Comment (jonathanlew) [inline src/software/geom/polygon.h:53]: we need some tests for this function in  polygon_test.cpp

**Review by jonathanlew — CHANGES_REQUESTED** [review]
nice work, just a few comments

File: src/software/ai/hl/stp/play/enemy_ball_placement_play_test.cpp [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    game_state.updateRefereeCommand(RefereeCommand::BALL_PLACEMENT_THEM);
+    game_state.setBallPlacementPoint(ball_placement_point);
+    setGameState(game_state);
+
+    std::vector<ValidationFunction> terminating_validation_functions = {
+        [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
+            while (world_ptr->getMostRecentTimestamp() < Timestamp::fromSeconds(9.5))
+            {
+                yield("Timestamp not at 9.5s");
+            }
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/enemy_ball_placement_play_test.cpp]: We need proper validation rather than waiting for 9.5s. There are other cases of this in this file

**Review by jonathanlew — CHANGES_REQUESTED** [review]

File: src/software/ai/hl/stp/play/enemy_ball_placement_play_test.cpp [hunk 6]
```diff
@@ -37,10 +37,11 @@ TEST_F(EnemyBallPlacementPlayTest, test_ball_placement_center)
 
     std::vector<ValidationFunction> terminating_validation_functions = {
         [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
-            while (world_ptr->getMostRecentTimestamp() < Timestamp::fromSeconds(9.5))
+            while (world_ptr->getMostRecentTimestamp() < Timestamp::fromSeconds(2))
             {
-                yield("Timestamp not at 9.5s");
+                yield("");
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/enemy_ball_placement_play_test.cpp]: this should yield something more useful

File: src/software/ai/hl/stp/play/enemy_ball_placement_play_test.cpp [hunk 7]
```diff
@@ -37,10 +37,11 @@ TEST_F(EnemyBallPlacementPlayTest, test_ball_placement_center)
...
             }
+            robotHalt(world_ptr, yield);
         }};
 
     std::vector<ValidationFunction> non_terminating_validation_functions = {};
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/enemy_ball_placement_play_test.cpp:47]: we should check that robots don't violate any conditions

File: src/software/gui/full_system/threaded_full_system_gui.cpp [hunk 8]
```diff
@@ -71,6 +71,7 @@ void ThreadedFullSystemGUI::createAndRunFullSystemGUI()
         sensor_msg_buffer, view_area_buffer, worlds_received_per_second_buffer,
         primitives_sent_per_second_buffer, mutable_thunderbots_config);
     full_system_gui->show();
+    full_system_gui->setWindowState(Qt::WindowMaximized);
```
Comment (jonathanlew) [inline src/software/gui/full_system/threaded_full_system_gui.cpp]: please remove, this will be done in another PR

File: src/software/ai/hl/stp/play/enemy_ball_placement_play_test.cpp (hunk 7 above)
Comment (PhilD71) [reply src/software/ai/hl/stp/play/enemy_ball_placement_play_test.cpp:47]: I'm going to add that tomorrow

**Review by jonathanlew — APPROVED** [review]
